<a href="https://colab.research.google.com/github/njaincode/python_for_data_science/blob/main/Sorting_and_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sorting and cleaning 
---



In order to effectively analyse a dataset, often we need to prepare it first. 
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* Sort by a single column e.g.
  * `df.sort_values("Make") `
* Sort by multiple columns e.g. 
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make 
* Sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted* 

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score 
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [1]:
import pandas as pd

def sort_ex1():
  excel_url = 'https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true'

  # Practice reading certain columns only
  df_hap_2015 = pd.read_excel(excel_url, usecols=["Happiness Rank", "Happiness Score", "Country"])
    
  # Set some display options when run
  pd.options.display.max_rows= 20
  pd.options.display.max_columns= 12

  # sort the data by Happiness Score in ascending order and display sorted table
  df_sorted_happy_score = df_hap_2015.sort_values(by="Happiness Score", ascending=True)
  print(df_sorted_happy_score)

sort_ex1()

         Country  Happiness Rank  Happiness Score
157         Togo             158            2.839
156      Burundi             157            2.905
155        Syria             156            3.006
154        Benin             155            3.340
153       Rwanda             154            3.465
..           ...             ...              ...
4         Canada               5            7.427
3         Norway               4            7.522
2        Denmark               3            7.527
1        Iceland               2            7.561
0    Switzerland               1            7.587

[158 rows x 3 columns]


### Exercise 2 - sort by multiple columns, display the first 5 rows 
---

1. Sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order 
2. Display the first 5 rows of sorted data 

**Test output**:  
Records 122, 127, 147, 100, 96

In [2]:
import pandas as pd

def sort_ex2():
  excel_url = 'https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true'

  # Practice reading certain columns only
  sort_col = ["Economy (GDP per Capita)", "Health (Life Expectancy)"]
  df_hap_2015 = pd.read_excel(excel_url, usecols=sort_col)
    
  #print(df_hap_2015.iloc[[122,127,147,100,96]])

  # Sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order
  # The test output needs columns sorted in this order and not in stat_col
  sort_col1 = ["Health (Life Expectancy)", "Economy (GDP per Capita)"]
  df_sorted_economy_health = df_hap_2015.sort_values(by=sort_col1).head(5)
  # Display the first 5 rows of sorted data
  print(df_sorted_economy_health)

sort_ex2()

     Economy (GDP per Capita)  Health (Life Expectancy)
122                   0.33024                   0.00000
127                   0.99355                   0.04776
147                   0.07850                   0.06699
100                   0.71206                   0.07566
96                    0.37545                   0.07612


### Exercise 3 - sorting in descending order 
---
 
Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
43, 7, 144, 0, 3 Country and Region columns


In [3]:
import pandas as pd

def sort_descending():
  excel_url = 'https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true'

  # Practice reading certain columns only
  get_col = ["Country", "Region", "Freedom", "Trust (Government Corruption)"]
  df_hap_2015 = pd.read_excel(excel_url, usecols=get_col)

  # Columns to display
  disp_col = ["Country", "Region"]
  # Columns used to sort
  sort_col = ["Freedom", "Trust (Government Corruption)"]
  disp_col1 = ["Country", "Freedom", "Trust (Government Corruption)"]
  
  # For debug
  #print(df_hap_2015.loc[[43,7,144,0,3], disp_col1])
  
  # Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows
  
  df_sorted_freedom_trust = df_hap_2015.sort_values(by=sort_col, ascending=[False, False])
  # Display the last 5 rows of sorted data (should show min values)
  # FIXME Test output does not match
  print(df_sorted_freedom_trust[disp_col].tail(5))

sort_descending()

                    Country                           Region
136                  Angola               Sub-Saharan Africa
117                   Sudan               Sub-Saharan Africa
95   Bosnia and Herzegovina       Central and Eastern Europe
101                  Greece                   Western Europe
111                    Iraq  Middle East and Northern Africa


# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values 

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values 
---

1. Read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv 
2. check if any NA values exist in the dataframe and print the result 
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071) 



In [44]:
import pandas as pd

def check_null_values():
  csv_url = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv'

  df_london_housing = pd.read_csv(csv_url)

  # 1. Check if any NA values exist in the dataframe and print the result
  # isnull is an alias of isna, so both can be used interchangably
  if (df_london_housing.isna().values.any() == True):
    print(f'Data frame contains NULL entries')
  else:
    print(f'Data frame does not contain any NULL entries')

  # 2. Use df.info() to see which columns have null entries
  # Used panda function instead
  print(df_london_housing.columns[df_london_housing.isna().any()].tolist())

  # isna().any() returns an index
  print(df_london_housing.columns[df_london_housing.isna().any()])

  # Different way to look at data
  print("No. of columns containing null values")
  # isna.any -> if NA in any cell
  print(len(df_london_housing.columns[df_london_housing.isna().any()]))

  print("No. of columns not containing null values")
  # notna.all -> No NA in any cell
  print(len(df_london_housing.columns[df_london_housing.notna().all()]))

  print("Total no. of columns in the dataframe")
  print(len(df_london_housing.columns))


check_null_values()


Data frame contains NULL entries
['median_salary', 'life_satisfaction', 'recycling_pct', 'population_size', 'number_of_jobs', 'area_size', 'no_of_houses']
Index(['median_salary', 'life_satisfaction', 'recycling_pct',
       'population_size', 'number_of_jobs', 'area_size', 'no_of_houses'],
      dtype='object')
No. of columns containing null values
7
No. of columns not containing null values
5
Total no. of columns in the dataframe
12


### Exercise 5 - remove null values 
---

1. Remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. Remove all NA values across whole dataframe 

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [55]:
import pandas as pd

def remove_null_values():
  csv_url = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv'

  df_london_housing = pd.read_csv(csv_url)

  # 1. Check if any NA values exist in life_satisfaction column
  if (df_london_housing["life_satisfaction"].isna().values.any() == True):
    print(f'life_satisfaction column contains NULL entries')

    print(f'Number of rows before dropna = {len(df_london_housing.index)}')

    df_nona_life_satisfaction = df_london_housing["life_satisfaction"].dropna()
    print(f'Number of rows after dropna = {len(df_nona_life_satisfaction.index)}')
    
  # 2. Remove all NA values across whole dataframe
  if (df_london_housing.isna().values.any() == True):
    print(f'Dataframe contains NULL entries')
    total_na_cells = df_london_housing.isna().sum()
    print(f'Total number of cells per col with NA = {total_na_cells}')

    df_london_housing.dropna(inplace=True)
    print(f'Number of rows after dropna on whole df = {len(df_london_housing.index)}')

    #total_na_cells_rows = df_london_housing.isna().sum(axis=1)
    #print(f'Total number of rows with NA = {total_na_cells_rows}')

remove_null_values()

life_satisfaction column contains NULL entries
Number of rows before dropna = 1071
Number of rows after dropna = 352
352
Dataframe contains NULL entries
Total number of cells with NA = code                   0
area                   0
date                   0
median_salary         22
life_satisfaction    719
mean_salary            0
recycling_pct        211
population_size       53
number_of_jobs       140
area_size            405
no_of_houses         405
borough_flag           0
dtype: int64
Number of rows after dropna on whole df = 267


## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])` 

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries 
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01* 

In [80]:
import pandas as pd

def remove_duplicates():
  csv_url = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv'

  df_london_housing = pd.read_csv(csv_url)

  # Remove duplicate area entries keeping first instance
  print(f'Number of rows before dropna = {len(df_london_housing["area"].index)}')
  print(df_london_housing[["area", "date"]])

  # Find number of duplicate entries in column area
  total_num_duplicates_in_area = len(df_london_housing['area']) -len(df_london_housing['area'].drop_duplicates())
  print(f'Total num of duplicates entries in area = {total_num_duplicates_in_area}')

  # Remove entries
  #df_london_housing.drop_duplicates(keep='first', inplace=True)
  #df = df_london_housing.drop_duplicates(subset= ["area"])
  df = df_london_housing["area"].drop_duplicates()
  print(f'Total num of unique entries in area = {len(df.index)}')
  print(df)

  # df_london_housing["area"].drop_dup.. with inplace=True will not work! Use subset
  df_london_housing.drop_duplicates(subset="area", keep='first', inplace=True)
  print(df_london_housing[["area", "date"]])
  #print(df_london_housing["area"])
  num_rows_in_area = len(df_london_housing["area"].index)
  print(num_rows_in_area)


remove_duplicates()

Number of rows before dropna = 1071
                      area        date
0           city of london  1999-12-01
1     barking and dagenham  1999-12-01
2                   barnet  1999-12-01
3                   bexley  1999-12-01
4                    brent  1999-12-01
...                    ...         ...
1066         great britain  2019-12-01
1067     england and wales  2019-12-01
1068      northern ireland  2019-12-01
1069              scotland  2019-12-01
1070                 wales  2019-12-01

[1071 rows x 2 columns]
Total num of duplicates entries in area = 1020
Total num of unique entries in area = 51
0           city of london
1     barking and dagenham
2                   barnet
3                   bexley
4                    brent
              ...         
46           great britain
47       england and wales
48        northern ireland
49                scotland
50                   wales
Name: area, Length: 51, dtype: object
                    area        date
0         c

# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

In [11]:
import pandas as pd

# Used when whole dataframe (each row) is passed
#def normalise_income(df):
#  if df['median_salary'] < 27441:
#    return 1
#  elif df['median_salary'] < 30932 :
#    return 2
#  else:
#    return 3


def normalise_income(val):
  if val < 27441:
    return 1
  elif val < 30932 :
    return 2
  else:
    return 3
 
def normalise_data_set():
  csv_url = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv'

  df_london_housing = pd.read_csv(csv_url)

  #1. Check if any NA values exist in life_satisfaction column
  if (df_london_housing["median_salary"].isna().values.any() == True):
    print(f' median_salary column contains NULL entries')

    print(f'Number of rows before dropna = {len(df_london_housing.index)}')

    # To keep things simple, removed nan and empty rows
    df_nona_salary = df_london_housing["median_salary"].dropna()
    print(f'Number of rows after dropna = {len(df_nona_salary.index)}')

    # Create a copy in case need to cross check
    df_nona_salary_orig = df_nona_salary
    #print(df_nona_salary.describe())

    # Rather than updating actual df, can it go into a copy?
    # Since df_nona_salary is a series, normalise_income has been changed accordingly
    df_nona_salary["median_salary"] = df_nona_salary.apply(normalise_income)
    print(df_nona_salary["median_salary"].head(10))
    print(df_nona_salary["median_salary"].describe())

normalise_data_set()

 median_salary column contains NULL entries
Number of rows before dropna = 1071
Number of rows after dropna = 1049
0    3
1    1
2    1
3    1
4    1
5    1
6    1
7    1
8    1
9    1
Name: median_salary, dtype: int64
count    1049.000000
mean        1.751192
std         0.830054
min         1.000000
25%         1.000000
50%         2.000000
75%         3.000000
max         3.000000
Name: median_salary, dtype: float64


### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

In [22]:
import pandas as pd

# Used when whole dataframe (each row) is passed
def normalise_col1(df, df_median, df_upper_quartile):
  if df['number_of_jobs'] < df_median:
    return 1
  elif df['number_of_jobs'] < df_upper_quartile:
    return 2
  else:
    return 3

def normalise_col(df):
  if df['number_of_jobs'] < 157000.0:
    return 1
  elif df['number_of_jobs'] < 2217000.0:
    return 2
  else:
    return 3

def normalise_job():
  csv_url = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv'

  df_london_housing = pd.read_csv(csv_url)

  #1. Check if any NA values exist in number_of_jobs column
  if (df_london_housing["number_of_jobs"].isna().values.any() == True):
    print(f'number_of_jobs column contains NULL entries')

    print(f'Number of rows before dropna = {len(df_london_housing.index)}')

    # To keep things simple, removed nan and empty rows
    df_london_housing["number_of_jobs"].dropna(inplace=True)
    
    # Create a copy in case need to cross check
    df_nona_jobs_orig = df_london_housing

    # Capture median, upper quartile and max (for debug)
    print(df_london_housing["number_of_jobs"].describe())

    job_median = df_london_housing["number_of_jobs"].median()
    job_max = df_london_housing["number_of_jobs"].max()
    job_75_per = df_london_housing["number_of_jobs"].quantile([0.75])
    print(f'median = {job_median} max = {job_max} job_max = {job_75_per}')

    # FIXME How to pass args to normalise fn?
    df_london_housing["number_of_jobs"] = df_london_housing.apply(normalise_col, axis=1)
    print(df_london_housing["number_of_jobs"].describe())


normalise_job()

number_of_jobs column contains NULL entries
Number of rows before dropna = 1071
count    9.310000e+02
mean     3.188095e+06
std      8.058302e+06
min      4.700000e+04
25%      9.450000e+04
50%      1.570000e+05
75%      2.217000e+06
max      3.575000e+07
Name: number_of_jobs, dtype: float64
median = 157000.0 max = 35750000.0 job_max = 0.75    2217000.0
Name: number_of_jobs, dtype: float64
count    1071.000000
mean        1.914099
std         0.880790
min         1.000000
25%         1.000000
50%         2.000000
75%         3.000000
max         3.000000
Name: number_of_jobs, dtype: float64


## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1 

In [29]:
import pandas as pd

def normalise_house_vol(val):
  if val < 102402.0:
    return 1
  elif val < 126276.0:
    return 2
  else:
    return 3
 
def normalise_house():
  csv_url = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv'

  df_london_housing = pd.read_csv(csv_url)

  #1. Check if any NA values exist in life_satisfaction column
  if (df_london_housing["no_of_houses"].isna().values.any() == True):
    print(f' no_of_houses column contains NULL entries')

    # To keep things simple, removed nan and empty rows
    df_nona_house_vol = df_london_housing["no_of_houses"].dropna()
    
    house_median = df_nona_house_vol.median()
    house_75_per = df_nona_house_vol.quantile([0.75])
    print(f'median = {house_median}  per_75 = {house_75_per}')

    # Normalise number of houses
    df_nona_house_vol["housing_volume"] = df_nona_house_vol.apply(normalise_house_vol)
    print(df_nona_house_vol["housing_volume"].head(10))
    print(df_nona_house_vol["housing_volume"].describe())

normalise_house()

 no_of_houses column contains NULL entries
Number of rows before dropna = 1071
count    6.660000e+02
mean     8.814682e+05
std      3.690376e+06
min      5.009000e+03
25%      8.763550e+04
50%      1.024020e+05
75%      1.262760e+05
max      2.417217e+07
Name: no_of_houses, dtype: float64
median = 102402.0  per_75 = 0.75    126276.0
Name: no_of_houses, dtype: float64
102    1
103    1
104    3
105    1
106    1
107    3
108    1
109    3
110    2
111    2
Name: no_of_houses, dtype: int64
count    666.000000
mean       1.750751
std        0.830458
min        1.000000
25%        1.000000
50%        1.500000
75%        2.750000
max        3.000000
Name: no_of_houses, dtype: float64


### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.


In [1]:
import pandas as pd

def normalise_area(df):
  if df["area_size"] < 4323.0:
    return 0
  else:
    return 1
 
def normalise_house():
  csv_url = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv'

  df_london_housing = pd.read_csv(csv_url)

  # To keep things simple, removed nan and empty rows
  df_london_housing.dropna(inplace=True, subset=["area_size"])
      
  area_median = df_london_housing["area_size"].median()
  print(f'median = {area_median}')

  # Normalise number of houses
  df_london_housing["area_size_normalised"] = df_london_housing.apply(normalise_area, axis=1)
  print(df_london_housing["area_size"].head(10))
  print(df_london_housing["area_size_normalised"].head(10))
  #print(df_london_housing["area_size_normalised"].describe())

normalise_house()

median = 4323.0
102      315.0
103     3780.0
104     8675.0
105     6429.0
106     4323.0
107    15013.0
108     2179.0
109     8650.0
110     5554.0
111     8220.0
Name: area_size, dtype: float64
102    0
103    0
104    1
105    1
106    1
107    1
108    0
109    1
110    1
111    1
Name: area_size_normalised, dtype: int64
